### PR TITLE
Patrick Kofod Mogensen: Making Optim.jl Great Again

### PR DESCRIPTION
**Pull request by pkofod**
Checklist for prospect students for Google Summer of Code 2016:

- [x] I contacted NumFOCUS mentors before writing my proposal
- [x] I showed you my contribution (it can be any form: proof-of-concept project idea, some sample code or just a link to your commits from other project)
- [x] I linked to my sample contribution from the proposal
- [ ] I linked to my opened issues in numfocus/gsoc repository from the proposal

--- end standard content ---
Att @mlubin 

I am a bit unsure about the last checklist item. I have no written detailed descriptions in the time line, as I think it is better to wait for general comments before detailed planning.

### CODE REVIEW

**Comment by mlubin** [comment]
This looks great to me.

@johnmyleswhite @timholy, I reached out to @pkofod to encourage this submission. Feel free to give feedback.

**Comment by mlubin** [comment]
CC @IainNZ @joehuchette

File: 2016/proposals/patrick-kofod-mogensen.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+### Unit testing
+The last area which needs better testing is testing each individual method in the package. A more detailed testing of the package itself will guard against future bugs, and might reveal current bugs.
+
+## Schedule of Deliverables
+
+### May 23th - June 27th
+
+- Documentation
+- Performance
+- JuliaCon (?)
```
Comment (mlubin) [inline 2016/proposals/patrick-kofod-mogensen.md]: Could be more explicit that you've submitted a proposal for a workshop

**Comment by mlubin** [comment]
CC @MikeInnes for crosstalk with Julia GSoC

File: 2016/proposals/patrick-kofod-mogensen.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
+# Making Optim.jl great again
```
Comment (IainNZ) [inline 2016/proposals/patrick-kofod-mogensen.md]: https://www.youtube.com/watch?v=2Tt04ZSlbZ0

File: 2016/proposals/patrick-kofod-mogensen.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+To improve what you might call coverage, I will extend the test set significantly. There are many sources available, for example from the CUTE(r) collection, Himmelblau (1972), and more. The will be tested at least at their canonical starting values, possibly also elsewhere.
+
+The performance tests will be performed with regards to different measures such as: function evaluations, gradient evaluations, cpu time, memory allocation (for large problems), and more. Standard measures such as means, percentiles, and typical ranks will be presented, but also [performance profiles](http://link.springer.com/article/10.1007%2Fs101070100263).
+
+### Improved algorithms
+I propose to improve at least two if the current algorithms as part of this GSoC project. The two I have in mind (Simulated Annealing and Nelder Mead) both have code available, and I will make sure that they are adapted, benchmarked, and merged if they provide improvements over the current situation. Obviously, the test set needs to be implemented to make this decision. Other possibilities might present themselves along the way.
```
Comment (IainNZ) [inline 2016/proposals/patrick-kofod-mogensen.md]: at least two **of**

File: 2016/proposals/patrick-kofod-mogensen.md (hunk 2 above)
Comment (pkofod) [inline 2016/proposals/patrick-kofod-mogensen.md]: Couldn't think of a better title, and thought I would cater to the American reviewers. Good thing I have more time to brain storm this particular part. Unless I misread your air horn, and it is one of the rare "I applaud this"-air horns.

Comment (IainNZ) [inline 2016/proposals/patrick-kofod-mogensen.md]: Was just thinking of this https://www.youtube.com/watch?v=w-McIdVuY88

Comment (pkofod) [inline 2016/proposals/patrick-kofod-mogensen.md]: Hadn't seen that one :)

**Comment by timholy** [comment]
Hard to find anything to complain about. Especially the title.

Perhaps it might be worth mentioning a little more concretely how you plan to extend the test suite: do you plan to interface in some way with https://github.com/JuliaOptimizers/CUTEst.jl? If not, how do you plan to port in the problems? (For problems in Himmelblau, how many are you talking about and is there any alternative to typing them in?) Can you give us a sense for the distribution of dimensionalities of problems you're hoping to add tests for? (I think right now all tests are probably < 10 dimensions, an important case but hardly representative of the range that people use in optimization.) 

With regards to "improved algorithms": what will you do if you discover ways to improve accuracy or performance on, e.g., high-dimensional problems but there turns out to be a cost for low-dimensional problems (or vice versa)? To state it in a way that's a little more sophisticated: it's likely that in comparing two implementations, one implementation might be better for some problems but worse for others. How do you make a decision in a way that acknowledges the fact that your choice of test suite may bias the statistics?

Like I said, it's an exciting proposal already. I'm not saying you have to address all of these issues, but these are things that occurred to me while reading your proposal.

**Comment by pkofod** [comment]
> Hard to find anything to complain about. Especially the title.

Thank you for taking your time to read the proposal.

> Perhaps it might be worth mentioning a little more concretely how you plan to extend the test suite:

Good idea. For the dimensionality, I had (in my mind) planned to address the current low dimension bias directly, in a first step simply by adding more higher dimensional problems (especially easy for problems with a dimensionality parameter `n`). Possibly report performance numbers on all tests, a small dimension subset, and a large dimension subset. For very large dimensional problems it might be useful to use something like data profiles instead of performance profiles (Moré and Wild, SIAM J. Optimization 2009). 

> With regards to "improved algorithms": what will you do if you discover ways to improve accuracy or performance on, e.g., high-dimensional problems but there turns out to be a cost for low-dimensional problems (or vice versa)?

I will have to think more about this, and possibly read more about this. Generally, I have some reading to do in this area.

> Like I said, it's an exciting proposal already. I'm not saying you have to address all of these issues, but these are things that occurred to me while reading your proposal.

Again, thank you. I will try to address your comments in the next version.

File: 2016/proposals/patrick-kofod-mogensen.md (hunk 2 above)
Comment (mlubin) [inline 2016/proposals/patrick-kofod-mogensen.md]: Well, may want to tone this down since I'm not sure how the google reviewers will like it

**Comment by mlubin** [comment]
@rgaiacs, when should @pkofod send in an official submission through the GSoC site?

**Comment by pkofod** [comment]
March 25 as far as I can see at https://github.com/numfocus/gsoc/blob/master/CONTRIBUTING-students.md#default-instructions @mlubin. I will change to a more neutral title. I will also fill in the time line stuff. Suggestions and/or comments are still welcome.

**Comment by pkofod** [comment]
Any comments @mlubin ? https://github.com/pkofod/gsoc/blob/master/2016/proposals/patrick-kofod-mogensen.pdf for a compiled version

File: 2016/proposals/patrick-kofod-mogensen.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+The first work on this project will be to improve the documentation. The rest of the project proposal either depend on it, or benefit from it. For me, it will also be a great opportunity get to know the current code even better.
+
+The user documentation work will coincide somewhat with the JuliaCon workshop preparation. As such, it will be the second this on my list, although the work on the documentation won't really finished before the project stops. The workshop will provide good feedback with regards to the improved documentation. It will also be a great source of examples to include, inspired by the work of the attendees.
+
+### June 28nd - August 14th
+
+With a solid documentation in place, and with the input from JuliaCon, a thorough performance system will be put in place. It is important spend a sufficient amount of time to get this right, such that the system can be used in the future.
+
+While algorithmic improvements to the mentioned algorithms will is tempting to choose next, unit tests will receive attention first. As mentioned above, some work will already have been done here, but this period will be a good time to take stock.
```
Comment (mlubin) [inline 2016/proposals/patrick-kofod-mogensen.md]: s/will is/will be/

**Comment by pbanaszkiewicz** [comment]
@pkofod I think you should remove the PDF from this PR and only include it in the GSOC website (as far as I understand, you have to add it as your final proposal before the deadline).

**Comment by mlubin** [comment]
@pbanaszkiewicz FYI @pkofod has already submitted the PDF through the GSOC website

**Comment by pbanaszkiewicz** [comment]
@mlubin sure, alright. In that case: @rgaiacs 1) will we merge students' proposals after the deadline? 2) Should we keep the PDF? I think `.md` is easier to read on GitHub.

**Comment by pkofod** [comment]
I've removed the pdf.

**Comment by rgaiacs** [comment]
@pkofod Thanks.

> 1) will we merge students' proposals after the deadline?

Doing it now and warning students that didn't submit their proposals on https://summerofcode.withgoogle.com/.

> 2) Should we keep the PDF? I think .md is easier to read on GitHub.

Just the Markdown file.

**Pull request merged**